enhancements/microshift/microshift-greenboot.md

```diff
@@ -0,0 +1,350 @@
+verification loop by repeating reboots up-to `GREENBOOT_MAX_BOOT_ATTEMPTS` times
+before attempting to rollback to the previous version of the software.
+
+```mermaid
+graph TD
+    A[Software vN] --> |Software Upgrade| B[Reboot Pending]
+    B --> |System Reboot| D{Validate Software vN+1}
+    D --> |Valid| E[Software vN+1]
+    D --> |Invalid| F[Verification Loop]
+    F --> |Up to Max Reboots|D
+    F --> |All Failed Attempts| G[Rollback]
+    G --> A
+```
+
+### API Extensions
+
+None
+
+### Implementation Details
+
+The MicroShift core RPM package will include the `microshift-greenboot-required.sh`
+health check script to validate that all the required MicroShift services are
+up and running. There will be no explicit dependency on the `greenboot` RPM.
+The script will be installed into the `/etc/greenboot/check/required.d` directory
+and it will not be executed during the system boot in case the `greenboot` package
+is not present.
+
+> If other scripts are present in the `/etc/greenboot/check/required.d` directory,
+> they will be executed by `greenboot` in the alphanumeric order.
+
+The script performs the following sanity checks:
+* Exit if the current user is not `root`
+* Exit if MicroShift service is not enabled
+* Wait for the MicroShift service to become active
+
+If the `greenboot-default-health-check` RPM subpackage is installed, it already
+includes [health check scripts](https://github.com/fedora-iot/greenboot#health-checks-included-with-subpackage-greenboot-default-health-checks)
+verifying that DNS and `ostree` services can be accessed.
+
+The health check script will implement validations for the MicroShift health
+endpoints and its services running as pods in the following namespaces:
+- openshift-ovn-kubernetes
+- openshift-service-ca
+- openshift-ingress
+- openshift-dns
+- openshift-storage
+
+> When pluggable CNI and CSI functionality is implemented, the `openshift-ovn-kubernetes`
+> and `openshift-storage` namespace name selection need to be conditional on
+> the loaded CNI and CSI plugins. The plugin developers may choose to add plugin-
+> specific health check scripts.
+
+Exiting the script with a non-zero status will have the boot declared as failed.
+Greenboot redirects all the script output to the system log, accessible via the
+`journalctl -u greenboot-healthcheck.service` command.
+
+|Validation                                           |Pass  |Fail  |
+|-----------------------------------------------------|------|------|
+|Check microshift.service is enabled                  |Next  |exit 0|
+|Wait for microshift.service to be active (!failed)   |Next  |exit 1|
+|Wait for Kubernetes API health endpoints to be OK    |Next  |exit 1|
+|Wait for any pods to start                           |Next  |exit 1|
+|For each core namespace, wait for images to be pulled|Next  |exit 1|
+|For each core namespace, wait for pods to be ready   |Next  |exit 1|
+|For each core namespace, check pods not restarting   |exit 0|exit 1|
```

Comment:
Why make this so fine-granular? What if instead we'd just implement the pattern of a `wait-for` subcommand that blocks on any/all of these conditions? It's something we probably also need for CI and other use cases (it's been asked so frequently). Also, it's easier to check programmatically when there's no `oc` on the host.

Reply:
I like the simplicity of that.

Reply:
Making this fine-granular allows for better understanding on what went wrong:
- I was already asked to split the image pulling part to signal a clear failure reason when networks are slower than needed
- The pod restart check is required in the end to make sure services are stable

The case of no `oc` on the host should not be considered, IMO. The relative size of that package is so small that not having it installed is causing no real size saving, while introducing a headache for developers.

Reply:
In this design document, it is important to be detailed about what conditions will be checked, so we should keep the table and make sure it includes everything necessary and it's in the right order, etc.

With https://github.com/openshift/microshift/pull/1280 we won't need a separate oc command. Since we already have an implementation queued up, and this is a development short cycle, I'm inclined to go ahead with that approach and see if we run into any issues. If we find that we need to reimplement part or all of the checks in go for reliability or other reasons, we can add a wait-for command to the binary in the future and make adjustments.

Reply:
I do agree it'd be useful for users to understand why the health-check failed, but wouldn't they already see that from the MicroShift logs? If we wanted to additionally expose a more explicit high-level status, we could implement that via [status notifications to systemd](https://unix.stackexchange.com/questions/212059/systemd-custom-status-message).

Reply:
I think that checking MicroShift logs is more of a task of developers for troubleshooting, because of the complexity of deduction on what went wrong - it requires knowledge of microshift components and internals. 

Sysadms may find it "easier" to have straightforward steps and reports from microshift greenboot script.

Reply:
I added a note about using `microshift get` instead of `oc get`